clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7430,6 +7430,212 @@ static void handleSYCLIntelFPGAMaxConcurrencyAttr(Sema &S, Decl *D,
   S.AddSYCLIntelFPGAMaxConcurrencyAttr(D, A, E);
 }
 
+// Checks if an expression is a valid filter list for an add_ir_attributes_*
+// attribute. Returns true if an error occured.
+static bool checkAddIRAttributesFilterListExpr(Expr *FilterListArg, Sema *S,
+                                               const AttributeCommonInfo &CI) {
+  assert(isa<InitListExpr>(FilterListArg));
+  const auto *FilterListE = cast<InitListExpr>(FilterListArg);
+  for (const Expr *FilterElemE : FilterListE->inits())
+    if (!isa<StringLiteral>(FilterElemE))
+      return S->Diag(FilterElemE->getBeginLoc(),
+                     diag::err_sycl_add_ir_attribute_invalid_filter)
+             << CI;
+  return false;
+}
+
+// Returns true if a type is either an array of char or a pointer to char.
+static bool isAddIRAttributesValidStringType(QualType T) {
+  return (T->isArrayType() || T->isPointerType()) &&
+         T->getPointeeOrArrayElementType()->isCharType();
+}
+
+// Checks if an expression is a valid attribute name for an add_ir_attributes_*
+// attribute. Returns true if an error occured.
+static bool checkAddIRAttributesNameExpr(Expr *NameArg, Sema *S,
+                                         const AttributeCommonInfo &CI) {
+  // Only strings and const char * are valid name arguments.
```

Comment:
Should we also be checking that the user spelled the LLVM IR attribute name properly instead of writing nonsense that can't be forwarded to LLVM? (I'm less worried about cases where someone puts nonsense on purpose and more worried about cases where there's a simple typo that we should alert the user to, perhaps even with a fixit.)

Reply:
Preferably CFE should be oblivious to valid IR attribute names to minimize the amount of changes required when we need new IR attributes.

Reply:
I think the FE will be oblivious to the IR attribute names; this should be driven by tablegen with a backend interface. e.g., at the LLVM level there should be an interface for testing whether the given name is valid. Then Clang can call into the LLVM function to check for validity. This interface could also offer a possibly-empty list of names as suggestions for typo correction.

I think this is important for catching typos. If it looks like it's a fair amount of work to support though (perhaps LLVM doesn't have such an interface yet and we'd need to invent one), it can be done in a follow-up, but I think it's important to do if possible.

Reply:
FWIW, I insist on the changing being made so that we diagnose this, but I don't insist that it be done as part of this review (it's fine to handle in a follow-up so long as the work gets done).